backend/src/apiserver/storage/run_store.go

```diff
@@ -350,6 +353,33 @@ func (s *RunStore) CreateRun(r *model.RunDetail) (*model.RunDetail, error) {
 }
 
 func (s *RunStore) UpdateRun(runID string, condition string, workflowRuntimeManifest string) (err error) {
+	tx, err := s.db.DB.Begin()
+	if err != nil {
+		return util.NewInternalServerError(err, "transaction creation failed")
+	}
+
+	// Lock the row for update, so we ensure no other update of the same run
```

Comment:
Why is it possible to add duplicated metadata? 

Updating metadata should idempotent since it is based on the ID of the workflow and POD. Is there something that prevents metadata updates to be idempotent?

Reply:
ml-metadata is not idempotent. It returns an id whenever we store a new artifact. That id is not based on the id of the pod or workflow.

Reply:
How will this approach work once the metadata store is independent from KFP? It won't be possible to use a transaction involving both databases at that point.

Reply:
I'm not sure we'll need transactions to be honest. Maybe we will, but we can solve that problem when we get to it. That question doesn't seem to affect what we're doing here right?